### PR TITLE
ED64P Saves and virtual writeback

### PR DESCRIPTION
**Pull request by ariahiro64**
This is the current status with the save implementation.

* Most sram games work (re2 is known not to save because it bleeds into sram data)
* All eeprom games save
* All save types work with some issues
* most fram games works (cubivore is a regression but haven't seen any other game effected)
* all emulators save fine gb64 savestates even work
* Clear unnecessary bloat (lib cart has defines i can't access so there's duplicates)
* Currently all games at least boot (cubivore crashes after game start)
* Memory alignment issue that plagued altra64 is absent
* uses settings api now so no ghost files.
TODO:
* test for more edge cases and (hopefully) figure out re2/cubavore

### CODE REVIEW

**Comment by ariahiro64** [comment]
Everything works
However....
ed64 saves have a different format and don't work in this menu at all. I think we should warn of their usage.

**Comment by ariahiro64** [comment]
Everything is functioning normally
Apparently there was a slight bug but its fixed.
Edit: bug found by Christopher Greiner

**Comment by ariahiro64** [comment]
Flashram doesnt work atm

**Comment by ariahiro64** [comment]
flashram is now working and the remaining defines that have libcart equivalents aren't exposed.

**Comment by ariahiro64** [comment]
This is the current status with the save implementation.

~~1) anti-piracy triggers for save related checks (needs discussion for auto patching)~~ 
2) most sram games work (re2 is known not to save because it bleeds into sram data)
3) all eeprom games save
4) for flashram games that actually saves reliably persist (altra64 had issues)
5) haven't switched to the settings api (yet)
6) clear unnecessary bloat (lib cart has defines i can't access so there's duplicates)
7) currently all games at least boot
8) memory alignment issue that plagued altra64 is absent

File: src/menu/cart_load.c [hunk 1]
```diff
@@ -174,7 +175,7 @@ cart_load_err_t cart_load_emulator (menu_t *menu, cart_load_emu_type_t emu_type,
             save_type = FLASHCART_SAVE_TYPE_FLASHRAM;
             break;
         case CART_LOAD_EMU_TYPE_GAMEBOY_COLOR:
-            path_push(path, "gbc.v64");
+            path_push(path, "gbc.v64");
```
Comment (networkfusion) [inline src/menu/cart_load.c]:
```suggestion
            path_push(path, "gbc.v64");
```

File: src/flashcart/ed64/ed64.c [hunk 2]
```diff
@@ -15,45 +15,79 @@
...
 // #include "../menu/settings.h"
 
-// // This is a trial hack before using the settings API.
-// #ifndef LAST_ROM_FILE_PATH
-// #define LAST_ROM_FILE_PATH "/ed64/last_rom.txt"
-// #endif
-
+// This is a trial hack before using the settings API.
+#ifndef LAST_SAVE_FILE_PATH
+#define LAST_SAVE_FILE_PATH strip_sd_prefix("/menu/last_rom.txt")
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]:
```suggestion
#define LAST_SAVE_FILE_PATH "/menu/last_rom.txt"
```

File: src/flashcart/ed64/ed64.c [hunk 3]
```diff
@@ -15,45 +15,79 @@
...
+    TCHAR lrp_path[255*32];
 
-    // FIXME: Retrive a config file from the SD card that might have been set.
-    // This should include the location of the ROM and its save type.
-    // Then, if it is valid, perform a save.
-    // FIL lrp_fil;
-    // UINT lrp_br;
+    // FIXME: use max supported length
+    
+    if (f_open(&lrp_fil, LAST_SAVE_FILE_PATH, FA_READ) != FR_OK) {
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c:37]:
```suggestion
    if (f_open(&lrp_fil, strip_sd_prefix(LAST_SAVE_FILE_PATH), FA_READ) != FR_OK) {
```

File: src/flashcart/ed64/ed64.c (hunk 2 above)
Comment (networkfusion) [reply src/flashcart/ed64/ed64.c]: Are you sure that the `strip_sd_prefix` is needed? it should not be... If it is, it should be used where required and not in the define.

File: src/flashcart/ed64/ed64.c [hunk 4]
```diff
@@ -15,45 +15,79 @@
...
 
 extern int ed_exit(void);
 
 static flashcart_err_t ed64_init (void) {
 
-    // TODO: partly already done, see https://github.com/DragonMinded/libdragon/blob/4ec469d26b6dc4e308caf3d5b86c2b340b708bbd/src/libcart/cart.c#L1064
-
-    // FIXME: Update firmware if needed.
-    // FIXME: Enable RTC if available.
+    if (file_exists(strip_sd_prefix("/menu/RESET"))) { // checks if reset
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]: To make this easier to maintain and change, use a define (as done with `LAST_SAVE_FILE_PATH`)

File: src/flashcart/ed64/ed64.c [hunk 5]
```diff
@@ -15,45 +15,79 @@
...
+    
+    f_unlink(strip_sd_prefix("/menu/RESET"));
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]: To make this easier to maintain and change, use a define (as done with LAST_SAVE_FILE_PATH)

File: src/flashcart/ed64/ed64.c [hunk 6]
```diff
@@ -15,45 +15,79 @@
...
-    // if (f_open(&lrp_fil, LAST_ROM_PATH_FILE, FA_READ) != FR_OK) {
-    //     return FLASHCART_ERR_LOAD;
-    // }
+    int lrp_size = f_size(&lrp_fil);
 
-    // if (f_read(&lrp_fil, lrp_path, 1024, &lrp_br) != FR_OK) {
-    //     f_close(&lrp_fil);
-    //     return FLASHCART_ERR_LOAD;
-    // }
+    if (f_read(&lrp_fil, lrp_path, lrp_size, &lrp_br) != FR_OK) {
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]: Given we know the lrp file size, we should really be allocating the TCHAR lrp_path size to that (so move the declaration). It will need to be one CHAR larger as it needs to take into account the null CHAR.

**Comment by networkfusion** [comment]
Nice work! 💯 

I have made a few comments that would improve a few things before getting to a state where I will merge into my branch...
Once that is done, we move into real review madness 😅 .

The fact that all of this stuff works better than on Altra64 OS is good to know.

Just note that it will be harder to get things merged to the N64FlashcartMenu as it is better to "try" and keep the code in a maintainable (and documented) state, but not put off people that mean they create their own release forks.

Hopefully the reviews so far have been fair 😁

File: src/flashcart/ed64/ed64.c [hunk 7]
```diff
@@ -15,45 +15,79 @@
...
+    UINT br;
+    uint8_t cartsave_data[KiB(128)];
+    // Older everdrives cant save during gameplay so we need to the reset method.
+
+    // find the path to last save
+    if (file_exists(strip_sd_prefix(lrp_path))) {
+
+    int save_size = file_get_size(strip_sd_prefix(lrp_path));
+
+    if ((f_open(&fil, strip_sd_prefix(lrp_path), FA_CREATE_ALWAYS | FA_READ | FA_WRITE)) != FR_OK) {
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]: If the lrp path size is too small, use the path_resize function to enlarge it.

File: src/flashcart/ed64/ed64.c [hunk 8]
```diff
@@ -175,42 +190,67 @@ static flashcart_err_t ed64_load_file (char *file_path, uint32_t rom_offset, uin
...
+
+    if (f_close(&lsp_fil) != FR_OK) {
+         return FLASHCART_ERR_LOAD;
+    }
+
+    FIL rsfil;
+
+    // this wries a 1 byte file as it only needs to exist to detect a reset
+
+    if (f_open(&rsfil, "/menu/RESET", FA_CREATE_ALWAYS ) != FR_OK) {
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]: To make this easier to maintain and change, use a define (as done with LAST_SAVE_FILE_PATH)

File: src/flashcart/ed64/ed64.c (hunk 2 above)
Comment (networkfusion) [reply src/flashcart/ed64/ed64.c]: Extra note, the `strip_sd_prefix` is used to "strip" the file path of `sd:` for which we do not use here (or some other places in this PR (but possibily not all).

File: src/flashcart/ed64/ed64.c [hunk 9]
```diff
@@ -15,45 +15,79 @@
...
+// This is a trial hack before using the settings API.
+#ifndef LAST_SAVE_FILE_PATH
+#define LAST_SAVE_FILE_PATH strip_sd_prefix("/menu/last_rom.txt")
+#endif
 
 extern int ed_exit(void);
 
 static flashcart_err_t ed64_init (void) {
 
-    // TODO: partly already done, see https://github.com/DragonMinded/libdragon/blob/4ec469d26b6dc4e308caf3d5b86c2b340b708bbd/src/libcart/cart.c#L1064
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c:28]: We need to keep these as are needed for other ED64 carts...

File: src/flashcart/ed64/ed64.c [hunk 10]
```diff
@@ -15,45 +15,79 @@
...
-
-    // FIXME: Update firmware if needed.
-    // FIXME: Enable RTC if available.
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]: We need to keep these as are needed for other ED64 carts...

File: src/flashcart/ed64/ed64.c [hunk 11]
```diff
@@ -175,42 +190,67 @@ static flashcart_err_t ed64_load_file (char *file_path, uint32_t rom_offset, uin
...
 
-    size_t save_size = f_size(&fil);
+    size_t save_size = file_get_size(strip_sd_prefix(save_path));
+    uint8_t cartsave_data[save_size];
 
-    if (f_read(&fil, address, save_size, &br) != FR_OK) {
-        f_close(&fil);
-        return FLASHCART_ERR_LOAD;
+    if (f_read(&fil, cartsave_data, save_size, &br) != FR_OK) {
+            f_close(&fil);
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]:
```suggestion
        f_close(&fil);
```

File: src/flashcart/ed64/ed64.c [hunk 12]
```diff
@@ -175,42 +190,67 @@ static flashcart_err_t ed64_load_file (char *file_path, uint32_t rom_offset, uin
...
+            return FLASHCART_ERR_LOAD;
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]:
```suggestion
        return FLASHCART_ERR_LOAD;
```

File: src/flashcart/ed64/ed64.c (hunk 11 above)
Comment (networkfusion) [reply src/flashcart/ed64/ed64.c]: Just an example, but make sure spacing is 4 chars (tab) spacing.

**Comment by ariahiro64** [comment]
ill do the changes tn

**Comment by ariahiro64** [comment]
updated the pinned comment

File: src/menu/cart_load.c [hunk 13]
```diff
@@ -1,5 +1,6 @@
 #include <string.h>
 
+#include <libcart/cart.h>
```
Comment (networkfusion) [inline src/menu/cart_load.c]: Pretty sure you dont need this?!

File: src/flashcart/ed64/ed64.c [hunk 14]
```diff
@@ -12,68 +13,135 @@
...
-// #ifndef LAST_ROM_FILE_PATH
-// #define LAST_ROM_FILE_PATH "/ed64/last_rom.txt"
-// #endif
-
+// This is a trial hack before using the settings API.
+#ifndef LAST_SAVE_FILE_PATH
+#define LAST_SAVE_FILE_PATH "/menu/last_rom.txt"
+#endif
+#ifndef RESET_CHECK_FILE_PATH
+#define RESET_CHECK_FILE_PATH "/menu/RESET"
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]:
```suggestion
#define RESET_CHECK_FILE_PATH "/menu/RESET.tmp"
```

File: src/flashcart/ed64/ed64.c [hunk 15]
```diff
@@ -12,68 +13,135 @@
...
+#endif
+#ifndef FLASHRAM_CHECK_FILE_PATH
+#define FLASHRAM_CHECK_FILE_PATH "/menu/FLASHRAM"
```
Comment (networkfusion) [inline src/flashcart/ed64/ed64.c]:
```suggestion
#define FLASHRAM_CHECK_FILE_PATH "/menu/FLASHRAM.tmp"
```

**Comment by ariahiro64** [comment]
this menu now has better compatibility than altra64. all flash ram games work. now i just need to figure out re2. gb64 even savestates correctly.

**Comment by Suprapote** [comment]
Hello, good job with your ed64 integration.
But i have to report that in the game cuvivore, when you press start the game shows a black screen and only sound the music. This game saves in Fram and on a previous comit these things don't happen and you can play it perfectly.

**Comment by ariahiro64** [comment]
image
seems fine to me

**Comment by ariahiro64** [comment]
oh
wait

**Comment by ariahiro64** [comment]
> Hello, good job with your ed64 integration. But i have to report that in the game cuvivore, when you press start the game shows a black screen and only sound the music. This game saves in fram and on a previous comit these things don't happen and you can play it perfectly.

im not sure whats going on. did it save before? ideally all games should work but it seems that re2 and this are edge cases. i also updated the above post for the current state of this pr.

**Comment by Suprapote** [comment]
Yes, it can save on one of the previous commits, but i can't tell you what was the commit, sorry

**Comment by ariahiro64** [comment]
im looking into rn

**Comment by ariahiro64** [comment]
so i got it to work.... once. i think it takes a while as when i got it to work i took a shower in between the black screen and returning

**Comment by ariahiro64** [comment]
i think it only works if you corrupt the file? after that all subsequent saves work. i just think ed64ps are cursed.

**Comment by networkfusion** [comment]
I am tempted to merge this into my branch now, if only to get it to a review state.

Given that most games (and more than Altra64) work, it is probably better to add any that dont to "known issues" for the moment.

**Comment by networkfusion** [comment]
Actually, I am going to merge this into my PR as LGTM.

I will probably make a few more changes there before commiting it to a new branch on the official repo. PR's for fixes (and a review against `main` will be generated from that.

**Comment by ariahiro64** [comment]
> Yes, it can save on one of the previous commits, but i can't tell you what was the commit, sorry

i got cubivore working in the new pull request
i also added the functionality to altra
#12
